viabel/optimization_diagnostics.py

```diff
@@ -40,45 +72,56 @@ def compute_R_hat_halfway(chains, interval=100, start=1000):
 
     return np.array(r_hats_halfway)
 
-def compute_R_hat_halfway_light(chains, warmup_fraction=0.5):
+def compute_R_hat_moving(chains, warmup_fraction=0.5):
+    """
+    Compute the split-R-hat for multiple chains over a moving window
+    - taking only the last warmup_fraction percentage of the iterates.
+
+    Parameters
+    ----------
+    chains : multi-dimensional array, shape=(n_chains, n_iters, n_dimensions)
+    warmup_fraction: percentage of the iterates: (1 -warmup_fraction)*n_iters
+    where n_iters is the current number of total iterates.
+
+    Returns
+    -------
+    r_hat_current : the r-hat value of the current window.
+    """
     n_chains, n_iters, K= chains.shape
     warmup = int(warmup_fraction*n_iters)
-    r_hat_current = compute_R_hat(chains, warmup=n_iters//2)[1]
+    r_hat_current = compute_R_hat(chains, warmup=warmup)[1]
     return r_hat_current
 
-def monte_carlo_se(iterate_chains, warmup=500):
-    chains = iterate_chains[:, warmup:, :]
-    n_chains, N_iters, K = chains.shape[0], chains.shape[1], chains.shape[2]
-    mcse_combined_list = np.zeros((N_iters,K))
-    print(chains.shape)
-    for i in range(1,N_iters):
-        chains_sub = chains[:, :i,:]
-        n_chains, n_iters, K = chains_sub.shape[0], chains_sub.shape[1], chains_sub.shape[2]
-        chains_flat = np.reshape(chains_sub, (n_chains*i, K))
-        variances = np.var(chains, ddof=1, axis=1)
-        variances_combined = np.var(chains_flat, ddof=1, axis=0)
-        mean_var_chains = np.mean(variances, axis=0)
-        mcse_per_chain = np.sqrt(variances / n_iters)
-        mcse_combined = np.sqrt(variances_combined/(i*n_chains))
-        mcse_combined_list[i] = mcse_combined
-
-    #print(mcse_per_chain.shape)
-    return mcse_per_chain, mcse_combined_list
 
 # compute mcmcse for a chain/array
-def monte_carlo_se2(iterate_chains, warmup=500, param_idx=0):
-    '''
-    compute monte carlo standard error for all chains and one parameter at a time.
-    :param iterate_chains:
-    :param warmup: warmup iterates
-    :return: array of mcse error for all chains for a particular variational parameter.
-    '''
+def monte_carlo_se_moving(iterate_chains, warmup=500, param_idx=0):
```

Comment:
in this function and other below, make warmup default to half of chain length or something similar. hardcoded value is not robust

Reply:
fixed.